lib/private/Installer.php

```
@@ -265,58 +235,148 @@ public static function updateAppByOCSId($ocsId) {
 	}
 
 	/**
-	 * @param array $data
-	 * @return array
+	 * Downloads an app and puts it into the app directory
+	 *
+	 * @param string $appId
+	 * @param AppFetcher $appFetcher
+	 * @param IClientService $clientService
+	 * @param ITempManager $tempManager
+	 * @param ILogger $logger
+	 *
+	 * @return bool Whether the installation was successful or not
 	 * @throws \Exception
 	 */
-	public static function downloadApp($data = array()) {
-		$l = \OC::$server->getL10N('lib');
-
-		if(!isset($data['source'])) {
-			throw new \Exception($l->t("No source specified when installing app"));
-		}
+	public function downloadApp($appId,
+								AppFetcher $appFetcher,
+								IClientService $clientService,
+								ITempManager $tempManager,
+								ILogger $logger) {
+		$appId = strtolower($appId);
+
+		$apps = $appFetcher->get();
+		foreach($apps as $app) {
+			if($app['id'] === $appId) {
+				// Verify if the certificate has been issued by the Nextcloud Code Authority CA
+				$x509 = new X509();
+				$x509->loadCA(file_get_contents(__DIR__ . '/../../resources/codesigning/root.crt'));
+				$x509->loadX509($app['certificate']);
+				if($x509->validateSignature() !== true) {
+					$logger->error(
+						sprintf(
+							'App with id %s has a certificate not issued by a trusted Code Signing Authority',
+							$appId
+						),
+						[
+							'app' => 'core',
+						]
+					);
+					return false;
```

Any reason for this? Should probably throw an exception with your logging string